File: python/inheritance_test.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals


class Muslim:

    def performsMuslimRitual(self):
        return


class MuslimFemale(Muslim):

    def performsMuslimRitual(self):
        print("Muslim girl performs Muslim ritual.")


class Hindu:

    def performsHinduRitual(self):
        return


class HinduFemale(Hindu):

    def performsHinduRitual(self):
        print("Hindu girl performs Hindu ritual.")


class MuslimRitual:

    def carryOutRitual(self, muslim):
        print("On with the Muslim rituals!")
        muslim.performsMuslimRitual()


class MuslimAdapter(Muslim):

    def __init__(self, h):
        self.hindu = h

    def performsMuslimRitual(self):
        self.hindu.performsHinduRitual()


if __name__ == '__main__':
    hinduGirl = HinduFemale()
    muslimGirl = MuslimFemale()
    muslimRitual = MuslimRitual()
    try:
        muslimRitual.carryOutRitual(hinduGirl)
    except AttributeError as exc:
        print(exc)
    adaptedHindu = MuslimAdapter(hinduGirl)

    muslimRitual.carryOutRitual(muslimGirl)
    muslimRitual.carryOutRitual(adaptedHindu)

File: cpp/inheritance_test.cpp
#include <iostream>
 
class Muslim {  // Abstract Target
//     virtual ~Muslim() = 0;
public:
    virtual void performsMuslimRitual() = 0;
};
 
class MuslimFemale : public Muslim {  // Concrete Target
    virtual void performsMuslimRitual() override {std::cout << "Muslim girl performs Muslim ritual." << std::endl;}
};
 
class Hindu {  // Abstract Adaptee
//     virtual ~Hindu() = 0;
public:
    virtual void performsHinduRitual() = 0;
};
 
class HinduFemale : public Hindu {  // Concrete Adaptee
    virtual void performsHinduRitual() override {std::cout << "Hindu girl performs Hindu ritual." << std::endl;}
};
 
class MuslimRitual {
public:
    void carryOutRitual (Muslim* muslim) {
        std::cout << "On with the Muslim rituals!" << std::endl;
        muslim->performsMuslimRitual();
    }
};
 
class MuslimAdapter : public Muslim {  // Adapter
    private:
        Hindu* hindu;
    public:
        MuslimAdapter (Hindu* h) : hindu(h) {}
        virtual void performsMuslimRitual() override {hindu->performsHinduRitual();}
};
 
int main() {  // Client code
    HinduFemale hinduGirl;
    MuslimFemale muslimGirl;
    MuslimRitual muslimRitual;
    //muslimRitual.carryOutRitual(&hinduGirl);  // Will not compile of course since the parameter must be of type Muslim*.
    MuslimAdapter adaptedHindu(&hinduGirl);  // hinduGirl has adapted to become a Muslim!
 
    muslimRitual.carryOutRitual(&muslimGirl);
    muslimRitual.carryOutRitual(&adaptedHindu);  // So now hinduGirl, in the form of adaptedHindu, participates in the muslimRitual!
                // Note that hinduGirl is carrying out her own type of ritual in muslimRitual though.
 
}
